Detectors/ITSMFT/ITS/reconstruction/include/ITSReconstruction/Clusterer.h

```diff
@@ -0,0 +1,54 @@
+/// \file Clusterer.h
+/// \brief Definition of the ITS cluster finder
+#ifndef ALICEO2_ITS_CLUSTERER_H
+#define ALICEO2_ITS_CLUSTERER_H
+
+#include <utility>
+#include <vector>
+
+#include "Rtypes.h"
+
+class TClonesArray;
+
+namespace o2
+{
+namespace ITS
+{
+  class PixelReader;
+  class Clusterer
+{
+ public:
+  Clusterer();
+  ~Clusterer() = default;
+
+  Clusterer(const Clusterer&) = delete;
+  Clusterer& operator=(const Clusterer&) = delete;
+
+  static void setPixelGeometry(Float_t px, Float_t pz, Float_t x0, Float_t z0) {
+    mPitchX=px; mPitchZ=pz; mX0=x0; mZ0=z0;
+  }
+  void process(PixelReader &r, TClonesArray &clusters);
+
+ private:
+  enum {kMaxRow=650}; //Anything larger than the real number of rows (512 for ALPIDE)
+  void initChip(UShort_t chipID, UShort_t row, UShort_t col, Int_t label);
+  void updateChip(UShort_t chipID, UShort_t row, UShort_t col, Int_t label);
+  void finishChip(TClonesArray &clusters);
+
+  Int_t mColumn1[kMaxRow+2];
+  Int_t mColumn2[kMaxRow+2];
+  Int_t *mCurr, *mPrev;
+  std::vector<std::vector<std::pair<UShort_t,UShort_t>>> mPreClusters;
```

Comment:
Hi, indeed, I missed that in ``updateChip`` only the mIndices are modified. But having digits referring to clusterID will cost either a ``sort`` or a search in the ``finishChip``. Also, current Ncl^2 loop will become Ndigits^2 loop.
I think it will be more efficient to use one ``vector<preCluster>`` with
```
struct preCluster {
   int firstDigitId;
  // eventual MC info data
}
```
and another ``vector<digit>`` with
```
struct digit {
  short row;
  short col;
  int nextDigitID;  // next digit of the cluster
}
```
The vectors in this case can be reused (no repetitive allocation) and adding a new cluster M will be
```
vecDig.emplace_back(digit(row,col,-1));  
vecClus.push_back(preCluster( vecDig.size()-1 ))
```
while adding digit to existing pre-cluster pCl will be 
```
vecDig.emplace_back(digit(row,col, pCl.firstDigitId));
pCl.firstDigitId = vecDig.size()-1;
```